File: backends/python/app/services/auth_service.py
from passlib.context import CryptContext
from app.models import User
from app.repositories.factory import get_user_repository
from app.auth.jwt import create_token

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")


async def register(email: str, password: str) -> str:
    repo = get_user_repository()
    existing = await repo.find_by_email(email)
    if existing:
        raise ValueError("email already registered")
    hashed = pwd_context.hash(password)
    user = User(email=email, password_hash=hashed)
    user = await repo.save(user)
    return create_token(user.id, user.email)


async def login(email: str, password: str) -> str:
    repo = get_user_repository()
    user = await repo.find_by_email(email)
    if not user or not pwd_context.verify(password, user.password_hash):
        raise ValueError("invalid email or password")
    return create_token(user.id, user.email)

File: backends/go/internal/service/auth_service.go
package service

import (
	"errors"
	"multi-lang-backend-go/internal/domain"
	"multi-lang-backend-go/internal/jwt"
	"multi-lang-backend-go/internal/repository"

	"golang.org/x/crypto/bcrypt"
)

var ErrEmailExists = errors.New("email already registered")
var ErrInvalidCredentials = errors.New("invalid email or password")

type AuthService struct {
	repo *repository.Factory
	jwt  *jwt.Manager
}

func NewAuthService(repo *repository.Factory, jwt *jwt.Manager) *AuthService {
	return &AuthService{repo: repo, jwt: jwt}
}

func (s *AuthService) Register(email, password string) (string, error) {
	userRepo := s.repo.UserRepository()
	existing, _ := userRepo.FindByEmail(email)
	if existing != nil {
		return "", ErrEmailExists
	}
	hash, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
	if err != nil {
		return "", err
	}
	user := &domain.User{Email: email, PasswordHash: string(hash)}
	if err := userRepo.Save(user); err != nil {
		return "", err
	}
	return s.jwt.Generate(user.ID, user.Email)
}

func (s *AuthService) Login(email, password string) (string, error) {
	user, err := s.repo.UserRepository().FindByEmail(email)
	if err != nil || user == nil {
		return "", ErrInvalidCredentials
	}
	if err := bcrypt.CompareHashAndPassword([]byte(user.PasswordHash), []byte(password)); err != nil {
		return "", ErrInvalidCredentials
	}
	return s.jwt.Generate(user.ID, user.Email)
}
